### DISCUSSION TITLE
Specifying optional outputs

### DISCUSSION THREAD

**Discussion by annaschultze** [opensafely/documentation#378] [Q&A] [answered]
I'm running a dofile which outputs graphs if some conditions are met (in this instance, frequency in the smallest histogram bin is >5). This happens to never be the case in the dummy data (and might never be the case in the real data), and therefore the outputs are not generated. 

This throws an error when running through the yaml as it cannot find any outputs matching the patterns I've specified. It makes sense that it cannot find it as they're not generated, but I'd like for the file to run irrespective of this, and for the outputs only to be saved when they are generated (i.e, they are not disclosive) 

Essentially I thought this was a quicker way than manually checking the files and choosing to only release those meeting redaction criteria. Is there a way around this? The closest idea I have so far is just outputting a blank image with the same name, but that seems a bit awkward. 

Link to [repo and branch](https://github.com/opensafely/covid-vaccine-neuro-research/tree/SCCS_code), it's the last action that's problematic.

**Comment by wjchulme** [comment, marked as answer]
Outputting extra files is a sensible and simple workaround. And I don't know of anything else you could do. But as you're using very permissive globs to specify the outputs of the action (`plots_S1: output/plots/S1*`, `plots_S2: output/plots/S2*`) you don't have to necessarily create a blank image. It could just be two blank text files `output/plots/S1.txt` and `output/plots/S2.txt`. Or maybe include info on which graphs met the condition or not, so at least it doesn't _feel_ like a workaround!

**Reply by annaschultze** [reply]
Thanks Will - it's helpful to know I wasn't going down an unnecessarily complex route. I like the idea of outputting a file with the info on which graphs meet the condition vs. not - agree that feels slightly more useful..! :)

**Reply by iaindillingham** [reply]
More generally, you could `touch` the output files before they are generated, so output files always exist in some form: either `touch`ed (zero bytes) or generated. This would avoid the error and avoid having to generate blank images. If you needed more information, then you could write a separate logfile.

With Python, you can `touch` a file like this:

```py
import pathlib
my_file = pathlib.Path("output/my_file.png")
my_file.touch()
```

R or Stata may have similar functionality. Alternatively, you could call `touch` on the shell.
